### ISSUE TITLE
New rule: Prefer .Count > 0 on Lists over .Any()

### ISSUE THREAD

**Issue by 0liver** [meziantou/Meziantou.Analyzer#360]
**Is your feature request related to a problem? Please describe.**
I'd like to use more performant code when it's easily available.

**Describe the solution you'd like**
I'd like an analyzer to warn me when invoking `.Any()` on `List<T>` variables.

**Describe alternatives you've considered**
Always use the better version by hand. Doesn't work well in a team.

**Additional context**
`List<T>.Count` should be significantly faster than `Enumerable.Any()` looking at what the latter does:

```c#
public static bool Any<TSource>(this IEnumerable<TSource> source)
{
  if (source == null)
    throw Error.ArgumentNull(nameof (source));
  using (IEnumerator<TSource> enumerator = source.GetEnumerator())
  {
    if (enumerator.MoveNext())
      return true;
  }
  return false;
}
```

**Comment by meziantou** [comment]
`Any` is a little bit smarter than the code you provide <https://source.dot.net/#System.Linq/System/Linq/AnyAll.cs,11>, so there is not much difference in term of performance.

But for people needing to write hot-path code, a new rule could be useful. If the collection implements `ICollection`, `ICollection<T>` or `IReadOnlyCollection<T>`, it could suggest to use `collection.Count > 0` instead of `collection.Any()`

**Comment by meziantou** [comment]
New rule added: [MA0112](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0112.md). Must be enabled explicitly as explain in the PR description https://github.com/meziantou/Meziantou.Analyzer/pull/361#issue-1345377216

**Comment by EngRajabi** [comment]
benchmark https://www.tabsoverspaces.com/233649-comparing-speed-of-count-0-and-any
The any command is sometimes faster

**Comment by meziantou** [comment]
With recent changes in .NET concerning interface casting, the difference should be lower (the bencmark is almost 5-year old). Anyway, the rule is implemented!

**Comment by 0liver** [comment]
> benchmark https://www.tabsoverspaces.com/233649-comparing-speed-of-count-0-and-any
The any command is sometimes faster

The call to `.Any()` is always faster than the call to the **LINQ method `.Count()`** (except for `Array`s), but always magnitudes slower than the collection's `.Count` or `.Length` property.

**Comment by 0liver** [comment]
Thanks for the fast help!

I forgot to mention that I was looking at a .NET Framework project, and a v4.5.1 one on top of that, where the implementation of `.Any()` is simply what I posted.

With the [latest version](https://source.dot.net/#System.Linq/System/Linq/AnyAll.cs,11) you posted, @meziantou, the incentive to replace `.Any()` with `.Count > 0` is much smaller, so for .NET (Core) code it wouldn't make sense to suggest any change. For .NET Fx projects there may be a benefit though, if used often.
